readme.md
# Xolphin API wrapper for Node.js
xolphin-api-js is a library which allows quick integration of the [Xolphin REST API](https://api.xolphin.com) in Node.js to automated ordering, issuance and installation of SSL Certificates.

## About Xolphin
[Xolphin](https://www.xolphin.nl/) is the largest supplier of [SSL Certificates](https://www.sslcertificaten.nl) and [Digital Signatures](https://www.digitalehandtekeningen.nl) in the Netherlands. Xolphin has
a professional team providing reliable support and rapid issuance of SSL Certificates at an affordable price from industry leading brands such as Comodo, GeoTrust, GlobalSign, Thawte and Symantec.

## Library installation

Library can be installed via [npm](http://npmjs.com/)

```
npm install xolphin-api-js
```

And updated via

```
npm update xolphin-api-js
```

## Usage

### Client initialization

```js
var xolphin = require('xolphin-api-js')

var client = new xolphin.Client('<username>', '<password>');
```

### Requests

#### Getting list of requests

```js
client.request.all(function(err, requests){
    requests.forEach(function(request){
        console.log(request.id, request._embedded.product.brand);
    });
});
```

#### Getting request by ID

```js
client.request.get(961992625, function(err, request){
    console.log(err, request._embedded.product.id);
});
```

#### Request certificate

```js
var ccr = client.request.create(24, 1, '<csr_string>', "EMAIL");
ccr.address = "Address";
ccr.approverFirstName = "FirstName";
ccr.approverLastName = "LastName";
ccr.approverPhone = "+12345678901";
ccr.approverEmail = "email@domain.com";
ccr.zipcode = "123456";
ccr.city = "City";
//currently available languages: en, de, fr, nl
ccr.language = "en";
ccr.company = "Company";
ccr.subjectAlternativeNames.push("test1.domain.com");
ccr.subjectAlternativeNames.push("test2.domain.com");
ccr.dcv.push({
    domain: "test1.domain.com",
    dcvType: "EMAIL",
    approverEmail: "email1.domain.com"
});

client.request.send(ccr, function(err, result) {
    console.log(err, result.id);
});
```

#### Create a note

```js
client.request.sendNote(1234, 'My message',function(err, result){
    console.log('Errors: '+err);
    console.log('Message: '+result.message);
});
```

#### Get list of notes

```js
client.request.getNotes(1234, function(err, result){
    result.forEach(function(note){
       console.log('Message: '+note.message);
    });
});
```

#### Send a "Comodo Subscriber Agreement" email

```js
//currently available languages: en, de, fr, nl
client.request.sendComodoSAEmail(1234, 'mail@example.com', 'en', function(err, result){
    console.log('Errors: '+err);
    console.log('Message: '+result.message);
});
```

#### Request an "Encryption Everywhere" certificate
```js
var request = client.request.createEE();
request.csr = "<csr_string>"
request.approverFirstName = "FirstName";
request.approverLastName = "LastName";
request.approverPhone = "+12345678901";
request.approverEmail = "email@domain.com";
request.validate = false;
request.subjectAlternativeNames.push("test1.domain.com");
request.subjectAlternativeNames.push("test2.domain.com");

client.request.sendEE(request,function(err, result){
    console.log(result);
});
```

### Certificate

#### Certificates list and expirations

```js
client.certificate.all(function(err, certificates){
    certificates.forEach(function(certificate){
        console.log(certificate.id, '-', (new Date(certificate.dateExpired).getTime() <= (new Date()).getTime()) );
    });
});
```

#### Download certificate

```js
var fs = require('fs');
client.certificate.download(961983489, 'CRT', function(err, certificate){
    fs.writeFileSync('cert.crt', certificate);
});
```

### Support

#### Products list

```js
client.support.products(function(err, products){
    products.forEach(function(product){
        console.log(product.id, product.brand);
    });
});
```

#### Decode CSR

```js
client.support.decodeCSR('<csr_string>', function(err, data){
    console.log(data.type, data.size);
});
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [1.6.0] - 2023-11-23
### Added
- approverRepresentative support
- more paramaters for schedule-validation-call

## [1.5.0] - 2017-03-22
### Added
- Encryption Everywhere support

lib/xolphin.js
// Generated by CoffeeScript 1.11.1
(function() {
  var Certificate, Client, Request, Support, async, querystring, r, requests,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  r = require('request');

  async = require('async');

  querystring = require('querystring');

  requests = require('./requests.js');

  Request = (function() {
    function Request(client) {
      this.client = client;
      this.sendComodoSAEmail = bind(this.sendComodoSAEmail, this);
      this.sendNote = bind(this.sendNote, this);
      this.getNotes = bind(this.getNotes, this);
      this.scheduleValidationCall = bind(this.scheduleValidationCall, this);
      this.retryDCV = bind(this.retryDCV, this);
      this.uploadDocument = bind(this.uploadDocument, this);
      this.get = bind(this.get, this);
      this.send = bind(this.send, this);
      this.create = bind(this.create, this);
      this.all = bind(this.all, this);
    }

    Request.prototype.all = function(callback) {
      requests = [];
      return this.client.get("requests", {
        page: 1
      }, (function(_this) {
        return function(err, result) {
          var page, pages;
          requests = result._embedded.requests;
          page = result.page;
          pages = result.pages;
          return async.until(function() {
            return pages === page;
          }, function(cb) {
            _this.client.get("requests", {
              page: page + 1
            }, function(err, result) {
              requests = requests.concat(result._embedded.requests);
              page = result.page;
              pages = result.pages;
              return cb(err);
            });
          }, function(err) {
            callback(err, requests);
          });
        };
      })(this));
    };

    Request.prototype.create = function(product, years, csr, dcvType) {
      return new requests.CreateCertificateRequest(product, years, csr, dcvType);
    };

    Request.prototype.send = function(request, callback) {
      return this.client.post("requests", request.toArray(), callback);
    };

    Request.prototype.createEE = function() {
      return new requests.CreateEERequest();
    };

    Request.prototype.sendEE = function(request, callback) {
      return this.client.post("requests/ee", request.toArray(), callback);
    };

    Request.prototype.get = function(id, callback) {
      return this.client.get("requests/" + id, {}, (function(_this) {
        return function(err, result) {
          return callback(err, result);
        };
      })(this));
    };

    Request.prototype.uploadDocument = function(id, document, description, callback) {
      return this.client.post("requests/" + id + "/upload-document", {
        document: document,
        description: description
      }, callback);
    };

    Request.prototype.retryDCV = function(id, domain, dcvType, email, callback) {
      if (email == null) {
        email = '';
      }
      return this.client.post("requests/" + id + "/retry-dcv", {
        domain: domain,
        dcvType: dcvType,
        email: email
      }, callback);
    };

    Request.prototype.scheduleValidationCall = function(id, dateTime, timezone, callback, phoneNumber, extensionNumber, email, comments, action, language) {
      var request;
      if (timezone == null) {
        timezone = "Europe/Amsterdam";
      }
      if (action == null) {
        action = "ScheduledCallback";
      }
      if (language == null) {
        language = "en_us";
      }
      request = {
        date: dateTime.toISOString().split('T')[0],
        time: dateTime.toISOString().split('T')[1].split('.')[0].substring(0, 5),
        timezone: timezone
      };
      if (phoneNumber !== null) {
        request['phoneNumber'] = phoneNumber;
      }
      if (extensionNumber !== null) {
        request['extensionNumber'] = extensionNumber;
      }
      if (email !== null) {
        request['email'] = email;
      }
      if (comments !== null) {
        request['comments'] = comments;
      }
      if (action !== 'ScheduledCallback') {
        request['action'] = action;
      }
      if (language !== 'en_us') {
        request['language'] = language;
      }
      return this.client.post("requests/" + id + "/schedule-validation-call", request, callback);
    };

    Request.prototype.getNotes = function(id, callback) {
      return this.client.get("requests/" + id + "/notes", {}, (function(_this) {
        return function(err, result) {
          return callback(err, result._embedded.notes);
        };
      })(this));
    };

    Request.prototype.sendNote = function(id, note, callback) {
      return this.client.post("requests/" + id + "/notes", {
        message: note
      }, (function(_this) {
        return function(err, result) {
          return callback(err, result);
        };
      })(this));
    };

    Request.prototype.sendComodoSAEmail = function(id, to, lang, callback) {
      return this.client.post("requests/" + id + "/sa", {
        sa_email: to,
        language: lang
      }, (function(_this) {
        return function(err, result) {
          return callback(err, result);
        };
      })(this));
    };

    return Request;

  })();

  Certificate = (function() {
    function Certificate(client) {
      this.client = client;
      this.cancel = bind(this.cancel, this);
      this.renew = bind(this.renew, this);
      this.reissue = bind(this.reissue, this);
      this.download = bind(this.download, this);
      this.get = bind(this.get, this);
      this.all = bind(this.all, this);
    }

    Certificate.prototype.all = function(callback) {
      var certificates;
      certificates = [];
      return this.client.get("certificates", {
        page: 1
      }, (function(_this) {
        return function(err, result) {
          var page, pages;
          certificates = result._embedded.certificates;
          page = result.page;
          pages = result.pages;
          return async.until(function() {
            return pages === page;
          }, function(cb) {
            _this.client.get("certificates", {
              page: page + 1
            }, function(err, result) {
              certificates = certificates.concat(result._embedded.certificates);
              page = result.page;
              pages = result.pages;
              return cb(err);
            });
          }, function(err) {
            callback(err, certificates);
          });
        };
      })(this));
    };

    Certificate.prototype.get = function(id, callback) {
      return this.client.get("certificates/" + id, {}, (function(_this) {
        return function(err, result) {
          return callback(err, result);
        };
      })(this));
    };

    Certificate.prototype.download = function(id, format, callback) {
      return this.client.download("certificates/" + id + "/download", {
        format: format
      }, callback);
    };

    Certificate.prototype.reissue = function(id, request, callback) {
      return this.client.post("certificates/" + id + "/reissue", request.toArray(), callback);
    };

    Certificate.prototype.renew = function(id, request, callback) {
      return this.client.post("certificates/" + id + "/renew", request.toArray(), callback);
    };

    Certificate.prototype.cancel = function(id, reason, revoke, callback) {
      return this.client.post("certificates/" + id + "/cancel", {
        reason: reason,
        revoke: revoke
      }, callback);
    };

    return Certificate;

  })();

  Support = (function() {
    function Support(client) {
      this.client = client;
      this.product = bind(this.product, this);
      this.products = bind(this.products, this);
      this.decodeCSR = bind(this.decodeCSR, this);
      this.approverEmailAddresses = bind(this.approverEmailAddresses, this);
    }

    Support.prototype.approverEmailAddresses = function(domain, callback) {
      return this.client.get("approver-email-addresses", {
        domain: domain
      }, callback);
    };

    Support.prototype.decodeCSR = function(csr, callback) {
      return this.client.post("decode-csr", {
        csr: csr
      }, callback);
    };

    Support.prototype.products = function(callback) {
      var products;
      products = [];
      return this.client.get("products", {
        page: 1
      }, (function(_this) {
        return function(err, result) {
          var page, pages;
          products = result._embedded.products;
          page = result.page;
          pages = result.pages;
          return async.until(function() {
            return pages === page;
          }, function(cb) {
            _this.client.get("products", {
              page: page + 1
            }, function(err, result) {
              products = products.concat(result._embedded.products);
              page = result.page;
              pages = result.pages;
              return cb(err);
            });
          }, function(err) {
            callback(err, products);
          });
        };
      })(this));
    };

    Support.prototype.product = function(id, callback) {
      return this.client.get("products/" + id, {}, (function(_this) {
        return function(err, result) {
          return callback(err, result);
        };
      })(this));
    };

    return Support;

  })();

  Client = (function() {
    function Client(username, password, test) {
      this.username = username;
      this.password = password;
      this.test = test;
      this._toObject = bind(this._toObject, this);
      this.post = bind(this.post, this);
      this.download = bind(this.download, this);
      this.get = bind(this.get, this);
      this.r = r.defaults({
        headers: {
          'Accept': 'application/json',
          'User-Agent': 'xolphin-api-js/1.6.0'
        }
      });
      if (this.test) {
        this.env = "@xolphin-public-api.loc/v1/";
      } else {
        this.env = "@api.xolphin.com/v1/";
      }
      this.baseUrl = "https://" + this.username + ":" + this.password + this.env;
      this.support = new Support(this);
      this.request = new Request(this);
      this.certificate = new Certificate(this);
    }

    Client.prototype.get = function(method, data, callback) {
      if (data == null) {
        data = {};
      }
      return this.r.get({
        url: "" + this.baseUrl + method + "?" + (querystring.stringify(data))
      }, (function(_this) {
        return function(e, r, b) {
          var ref;
          b = _this._toObject(b);
          if (e != null) {
            if (typeof callback === 'function') {
              return callback(true, e, b);
            }
          } else {
            if ((200 <= (ref = r.statusCode) && ref < 300)) {
              if (!b.hasOwnProperty('message') && !b.hasOwnProperty('errors')) {
                if (typeof callback === 'function') {
                  return callback(false, b, b);
                }
              } else {
                if (typeof callback === 'function') {
                  return callback(true, b.message, b);
                }
              }
            } else {
              if (typeof callback === 'function') {
                return callback(true, b, b);
              }
            }
          }
        };
      })(this));
    };

    Client.prototype.download = function(method, data, callback) {
      if (data == null) {
        data = {};
      }
      return this.r.get({
        url: "" + this.baseUrl + method + "?" + (querystring.stringify(data))
      }, (function(_this) {
        return function(e, r, b) {
          var ref;
          b = _this._toObject(b);
          if (e != null) {
            if (typeof callback === 'function') {
              return callback(true, e, b);
            }
          } else {
            if ((200 <= (ref = r.statusCode) && ref < 300)) {
              if (typeof callback === 'function') {
                return callback(false, b, b);
              }
            } else {
              if (typeof callback === 'function') {
                return callback(true, b, b);
              }
            }
          }
        };
      })(this));
    };

    Client.prototype.post = function(method, data, callback) {
      if (data == null) {
        data = {};
      }
      return this.r.post({
        url: "" + this.baseUrl + method,
        formData: data
      }, (function(_this) {
        return function(e, r, b) {
          var ref;
          b = _this._toObject(b);
          if (e != null) {
            if (typeof callback === 'function') {
              return callback(true, e, b);
            }
          } else {
            if ((200 <= (ref = r.statusCode) && ref < 300)) {
              if (!b.hasOwnProperty('errors')) {
                if (typeof callback === 'function') {
                  return callback(false, b, b);
                }
              } else {
                if (typeof callback === 'function') {
                  return callback(true, b.message, b);
                }
              }
            } else {
              if (typeof callback === 'function') {
                return callback(true, b, b);
              }
            }
          }
        };
      })(this));
    };

    Client.prototype._toObject = function(json) {
      try {
        JSON.parse(json);
      } catch (error) {
        return json;
      }
      return JSON.parse(json);
    };

    return Client;

  })();

  module.exports = {
    Client: Client
  };

}).call(this);


lib/requests.js
// Generated by CoffeeScript 1.11.1
(function() {
  var CreateCertificateRequest, CreateEERequest, ReissueCertificateRequest, RenewCertificateRequest,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  CreateCertificateRequest = (function() {
    function CreateCertificateRequest(product, years, csr, dcvType) {
      this.product = product;
      this.years = years;
      this.csr = csr;
      this.dcvType = dcvType;
      this.toArray = bind(this.toArray, this);
      this.subjectAlternativeNames = [];
      this.dcv = [];
      this.certenroll_email = '';
      this.company = '';
      this.department = '';
      this.address = '';
      this.zipcode = '';
      this.city = '';
      this.province = '';
      this.country = '';
      this.approverFirstName = '';
      this.approverLastName = '';
      this.approverEmail = '';
      this.approverPhone = '';
      this.approverRepresentativePosition = '';
      this.approverRepresentativeFirstName = '';
      this.approverRepresentativeLastName = '';
      this.approverRepresentativeEmail = '';
      this.approverRepresentativePhone = '';
      this.kvk = '';
      this.reference = '';
      this.language = '';
    }

    CreateCertificateRequest.prototype.toArray = function() {
      var result;
      result = {
        product: this.product,
        years: this.years,
        csr: this.csr,
        dcvType: this.dcvType
      };
      if (this.subjectAlternativeNames.length > 0) {
        result['subjectAlternativeNames'] = this.subjectAlternativeNames.join(',');
      }
      if (this.dcv.length > 0) {
        result['dcv'] = JSON.stringify(this.dcv);
      }
      if (this.certenroll_email !== '') {
        result['certenroll_email'] = this.certenroll_email;
      }
      if (this.company !== '') {
        result['company'] = this.company;
      }
      if (this.department !== '') {
        result['department'] = this.department;
      }
      if (this.address !== '') {
        result['address'] = this.address;
      }
      if (this.zipcode !== '') {
        result['zipcode'] = this.zipcode;
      }
      if (this.city !== '') {
        result['city'] = this.city;
      }
      if (this.province !== '') {
        result['province'] = this.province;
      }
      if (this.country !== '') {
        result['country'] = this.country;
      }
      if (this.approverFirstName !== '') {
        result['approverFirstName'] = this.approverFirstName;
      }
      if (this.approverLastName !== '') {
        result['approverLastName'] = this.approverLastName;
      }
      if (this.approverEmail !== '') {
        result['approverEmail'] = this.approverEmail;
      }
      if (this.approverPhone !== '') {
        result['approverPhone'] = this.approverPhone;
      }
      if (this.approverRepresentativePosition !== '') {
        result['approverRepresentativePosition'] = this.approverRepresentativePosition;
      }
      if (this.approverRepresentativeFirstName !== '') {
        result['approverRepresentativeFirstName'] = this.approverRepresentativeFirstName;
      }
      if (this.approverRepresentativeLastName !== '') {
        result['approverRepresentativeLastName'] = this.approverRepresentativeLastName;
      }
      if (this.approverRepresentativeEmail !== '') {
        result['approverRepresentativeEmail'] = this.approverRepresentativeEmail;
      }
      if (this.approverRepresentativePhone !== '') {
        result['approverRepresentativePhone'] = this.approverRepresentativePhone;
      }
      if (this.kvk !== '') {
        result['kvk'] = this.kvk;
      }
      if (this.reference !== '') {
        result['reference'] = this.reference;
      }
      if (this.language !== '') {
        result['language'] = this.language;
      }
      return result;
    };

    return CreateCertificateRequest;

  })();

  CreateEERequest = (function() {
    function CreateEERequest() {
      this.toArray = bind(this.toArray, this);
      this.csr = '';
      this.subjectAlternativeNames = [];
      this.approverRepresentativePosition = '';
      this.approverRepresentativeFirstName = '';
      this.approverRepresentativeLastName = '';
      this.approverRepresentativeEmail = '';
      this.approverRepresentativePhone = '';
      this.approverFirstName = '';
      this.approverLastName = '';
      this.approverEmail = '';
      this.approverPhone = '';
      this.dcvType = 'FILE';
      this.validate = false;
    }

    CreateEERequest.prototype.toArray = function() {
      var result;
      result = {};
      if (this.subjectAlternativeNames.length > 0) {
        result['subjectAlternativeNames'] = this.subjectAlternativeNames.join(',');
      }
      result = {
        csr: this.csr,
        approverFirstName: this.approverFirstName,
        approverLastName: this.approverLastName,
        approverEmail: this.approverEmail,
        approverPhone: this.approverPhone,
        dcvType: this.dcvType,
        validate: this.validate ? 1 : 0
      };
      if (this.approverRepresentativePosition !== '') {
        result['approverRepresentativePosition'] = this.approverRepresentativePosition;
      }
      if (this.approverRepresentativeFirstName !== '') {
        result['approverRepresentativeFirstName'] = this.approverRepresentativeFirstName;
      }
      if (this.approverRepresentativeLastName !== '') {
        result['approverRepresentativeLastName'] = this.approverRepresentativeLastName;
      }
      if (this.approverRepresentativeEmail !== '') {
        result['approverRepresentativeEmail'] = this.approverRepresentativeEmail;
      }
      if (this.approverRepresentativePhone !== '') {
        result['approverRepresentativePhone'] = this.approverRepresentativePhone;
      }
      return result;
    };

    return CreateEERequest;

  })();

  ReissueCertificateRequest = (function() {
    function ReissueCertificateRequest(csr, dcvType) {
      this.csr = csr;
      this.dcvType = dcvType;
      this.toArray = bind(this.toArray, this);
      this.subjectAlternativeNames = [];
      this.dcv = [];
      this.certenroll_email = '';
      this.company = '';
      this.department = '';
      this.address = '';
      this.zipcode = '';
      this.city = '';
      this.province = '';
      this.country = '';
      this.approverFirstName = '';
      this.approverLastName = '';
      this.approverEmail = '';
      this.approverPhone = '';
      this.approverRepresentativePosition = '';
      this.approverRepresentativeFirstName = '';
      this.approverRepresentativeLastName = '';
      this.approverRepresentativeEmail = '';
      this.approverRepresentativePhone = '';
      this.kvk = '';
      this.reference = '';
      this.language = '';
    }

    ReissueCertificateRequest.prototype.toArray = function() {
      var result;
      result = {
        csr: this.csr,
        dcvType: this.dcvType
      };
      if (this.subjectAlternativeNames.length > 0) {
        result['subjectAlternativeNames'] = this.subjectAlternativeNames.join(',');
      }
      if (this.dcv.length > 0) {
        result['dcv'] = JSON.stringify(this.dcv);
      }
      if (this.certenroll_email !== '') {
        result['certenroll_email'] = this.certenroll_email;
      }
      if (this.company !== '') {
        result['company'] = this.company;
      }
      if (this.department !== '') {
        result['department'] = this.department;
      }
      if (this.address !== '') {
        result['address'] = this.address;
      }
      if (this.zipcode !== '') {
        result['zipcode'] = this.zipcode;
      }
      if (this.city !== '') {
        result['city'] = this.city;
      }
      if (this.province !== '') {
        result['province'] = this.province;
      }
      if (this.country !== '') {
        result['country'] = this.country;
      }
      if (this.approverFirstName !== '') {
        result['approverFirstName'] = this.approverFirstName;
      }
      if (this.approverLastName !== '') {
        result['approverLastName'] = this.approverLastName;
      }
      if (this.approverEmail !== '') {
        result['approverEmail'] = this.approverEmail;
      }
      if (this.approverPhone !== '') {
        result['approverPhone'] = this.approverPhone;
      }
      if (this.approverRepresentativePosition !== '') {
        result['approverRepresentativePosition'] = this.approverRepresentativePosition;
      }
      if (this.approverRepresentativeFirstName !== '') {
        result['approverRepresentativeFirstName'] = this.approverRepresentativeFirstName;
      }
      if (this.approverRepresentativeLastName !== '') {
        result['approverRepresentativeLastName'] = this.approverRepresentativeLastName;
      }
      if (this.approverRepresentativeEmail !== '') {
        result['approverRepresentativeEmail'] = this.approverRepresentativeEmail;
      }
      if (this.approverRepresentativePhone !== '') {
        result['approverRepresentativePhone'] = this.approverRepresentativePhone;
      }
      if (this.kvk !== '') {
        result['kvk'] = this.kvk;
      }
      if (this.reference !== '') {
        result['reference'] = this.reference;
      }
      if (this.language !== '') {
        result['language'] = this.language;
      }
      return result;
    };

    return ReissueCertificateRequest;

  })();

  RenewCertificateRequest = (function() {
    function RenewCertificateRequest(product, years, csr, dcvType) {
      this.product = product;
      this.years = years;
      this.csr = csr;
      this.dcvType = dcvType;
      this.toArray = bind(this.toArray, this);
      this.subjectAlternativeNames = [];
      this.dcv = [];
      this.certenroll_email = '';
      this.company = '';
      this.department = '';
      this.address = '';
      this.zipcode = '';
      this.city = '';
      this.province = '';
      this.country = '';
      this.approverFirstName = '';
      this.approverLastName = '';
      this.approverEmail = '';
      this.approverPhone = '';
      this.approverRepresentativePosition = '';
      this.approverRepresentativeFirstName = '';
      this.approverRepresentativeLastName = '';
      this.approverRepresentativeEmail = '';
      this.approverRepresentativePhone = '';
      this.kvk = '';
      this.reference = '';
      this.language = '';
    }

    RenewCertificateRequest.prototype.toArray = function() {
      var result;
      result = {
        product: this.product,
        years: this.years,
        csr: this.csr,
        dcvType: this.dcvType
      };
      if (this.subjectAlternativeNames.length > 0) {
        result['subjectAlternativeNames'] = this.subjectAlternativeNames.join(',');
      }
      if (this.dcv.length > 0) {
        result['dcv'] = JSON.stringify(this.dcv);
      }
      if (this.certenroll_email !== '') {
        result['certenroll_email'] = this.certenroll_email;
      }
      if (this.company !== '') {
        result['company'] = this.company;
      }
      if (this.department !== '') {
        result['department'] = this.department;
      }
      if (this.address !== '') {
        result['address'] = this.address;
      }
      if (this.zipcode !== '') {
        result['zipcode'] = this.zipcode;
      }
      if (this.city !== '') {
        result['city'] = this.city;
      }
      if (this.province !== '') {
        result['province'] = this.province;
      }
      if (this.country !== '') {
        result['country'] = this.country;
      }
      if (this.approverFirstName !== '') {
        result['approverFirstName'] = this.approverFirstName;
      }
      if (this.approverLastName !== '') {
        result['approverLastName'] = this.approverLastName;
      }
      if (this.approverEmail !== '') {
        result['approverEmail'] = this.approverEmail;
      }
      if (this.approverPhone !== '') {
        result['approverPhone'] = this.approverPhone;
      }
      if (this.approverRepresentativePosition !== '') {
        result['approverRepresentativePosition'] = this.approverRepresentativePosition;
      }
      if (this.approverRepresentativeFirstName !== '') {
        result['approverRepresentativeFirstName'] = this.approverRepresentativeFirstName;
      }
      if (this.approverRepresentativeLastName !== '') {
        result['approverRepresentativeLastName'] = this.approverRepresentativeLastName;
      }
      if (this.approverRepresentativeEmail !== '') {
        result['approverRepresentativeEmail'] = this.approverRepresentativeEmail;
      }
      if (this.approverRepresentativePhone !== '') {
        result['approverRepresentativePhone'] = this.approverRepresentativePhone;
      }
      if (this.kvk !== '') {
        result['kvk'] = this.kvk;
      }
      if (this.reference !== '') {
        result['reference'] = this.reference;
      }
      if (this.language !== '') {
        result['language'] = this.language;
      }
      return result;
    };

    return RenewCertificateRequest;

  })();

  module.exports = {
    CreateCertificateRequest: CreateCertificateRequest,
    CreateEERequest: CreateEERequest,
    ReissueCertificateRequest: ReissueCertificateRequest,
    RenewCertificateRequest: RenewCertificateRequest
  };

}).call(this);


examples/Request.Create.EE.js
// example for creating a request
(function(){
    var Xolphin = require('../lib/xolphin.js');
    var client = new Xolphin.Client('nikita@xolphin.nl', 'test1234', true);

    var request = client.request.createEE();

    request.csr = '-----BEGIN CERTIFICATE REQUEST-----\
    MIICuDCCAaACADB0MQswCQYDVQQGEwJOTDEWMBQGA1UEAxMNd3d3LnNzbGRldi5u\
    bDEWMBQGA1UEBxMNSGVlcmh1Z293YWFyZDEQMA4GA1UEChMHWG9scGhpbjEWMBQG\
    A1UECBMNTm9vcmQtSG9sbGFuZDELMAkGA1UECxMCSVQwggEiMA0GCSqGSIb3DQEB\
    AQUAA4IBDwAwggEKAoIBAQDkTFMNKhWObiUH1xDeRcqYiROKvJZEI7njDlQbXeVC\
    qSaBSBClJD98OSgE0+arnVyJ5Fnjkovgb9aC8HbWYbiD1tbWD+hKC1wPODEvknl7\
    jfRpO36cIglIhftHn5p0k+f86uS2Nz1yr+5qwAEQ7C0UQZxMuBbzK6Uv8m0TbAVV\
    CQ+i+uJfuXSsAT8LZbyoJsu50ZySG3uEhCSISh1cS/t/M2INHbXfzGA3GWH2FXCZ\
    Qd9eLSBAVEanIr3TlRpNU3f0IDwJpm56BVTNtNL7GI2NilfUF9oSo9PSlPbAkPsN\
    BrVdOiZgiOEO74rXE8fjA5Zm8uic6MyFRo4FuTUBWubrAgMBAAGgADANBgkqhkiG\
    9w0BAQsFAAOCAQEAHuB6643oI7oPw5SbFQAcfTfQbyRTXNqwYOO8YRcCkDIu7aX9\
    pupGvf+cSa5IDcZ6Dz22/Khp5rYlcutAtEN7d8MNBwBlcVzbWL+DuTU5ZWAxcRi4\
    nmiw/C+RxFF+WsUcjwb+dNrTYkz03t7voMg/0NS7RRU/oTu8heDDcH+Ffam1bZJQ\
    zi21dz/AsjG9Jc6GJqZs0ImEwCQlBhJYnXPj3FB8U/mzSWSfq502fdtmQrvIgHmI\
    RLuxVO9QX0YLQ0ew1x2z+eFfBsfSI1+DkF4+5TrfJP6jQyHBR+VDdGYBO2d8rBw7\
    ITIJC9tt2F4GjaPAI1xY9eoq/QfZxzNoNufMuw==\
    -----END CERTIFICATE REQUEST-----'
    request.approverFirstName = "FirstName";
    request.approverLastName = "LastName";
    request.approverPhone = "+12345678901";
    request.approverEmail = "email@domain.com";
    request.validate = false;
    request.subjectAlternativeNames.push("test1.domain.com");
    request.subjectAlternativeNames.push("test2.domain.com");

    client.request.sendEE(request,function(err, result){
        console.log(result);
        console.log('Error: '+ err);
        console.log('Errors: ');
        console.log(result.errors)
    });

}).call(this);



examples/Request.Create.js
// example for creating a request
(function(){
    var Xolphin = require('../lib/xolphin.js');
    var client = new Xolphin.Client('fake_login@xolphin.api', 'Sup3rSecre7P@s$w0rdForThe@p1', true);

    var request = client.request.create(24, 1, '-----BEGIN CERTIFICATE REQUEST-----\
    MIICuDCCAaACADB0MQswCQYDVQQGEwJOTDEWMBQGA1UEAxMNd3d3LnNzbGRldi5u\
    bDEWMBQGA1UEBxMNSGVlcmh1Z293YWFyZDEQMA4GA1UEChMHWG9scGhpbjEWMBQG\
    A1UECBMNTm9vcmQtSG9sbGFuZDELMAkGA1UECxMCSVQwggEiMA0GCSqGSIb3DQEB\
    AQUAA4IBDwAwggEKAoIBAQDkTFMNKhWObiUH1xDeRcqYiROKvJZEI7njDlQbXeVC\
    qSaBSBClJD98OSgE0+arnVyJ5Fnjkovgb9aC8HbWYbiD1tbWD+hKC1wPODEvknl7\
    jfRpO36cIglIhftHn5p0k+f86uS2Nz1yr+5qwAEQ7C0UQZxMuBbzK6Uv8m0TbAVV\
    CQ+i+uJfuXSsAT8LZbyoJsu50ZySG3uEhCSISh1cS/t/M2INHbXfzGA3GWH2FXCZ\
    Qd9eLSBAVEanIr3TlRpNU3f0IDwJpm56BVTNtNL7GI2NilfUF9oSo9PSlPbAkPsN\
    BrVdOiZgiOEO74rXE8fjA5Zm8uic6MyFRo4FuTUBWubrAgMBAAGgADANBgkqhkiG\
    9w0BAQsFAAOCAQEAHuB6643oI7oPw5SbFQAcfTfQbyRTXNqwYOO8YRcCkDIu7aX9\
    pupGvf+cSa5IDcZ6Dz22/Khp5rYlcutAtEN7d8MNBwBlcVzbWL+DuTU5ZWAxcRi4\
    nmiw/C+RxFF+WsUcjwb+dNrTYkz03t7voMg/0NS7RRU/oTu8heDDcH+Ffam1bZJQ\
    zi21dz/AsjG9Jc6GJqZs0ImEwCQlBhJYnXPj3FB8U/mzSWSfq502fdtmQrvIgHmI\
    RLuxVO9QX0YLQ0ew1x2z+eFfBsfSI1+DkF4+5TrfJP6jQyHBR+VDdGYBO2d8rBw7\
    ITIJC9tt2F4GjaPAI1xY9eoq/QfZxzNoNufMuw==\
    -----END CERTIFICATE REQUEST-----', "EMAIL");

    request.address = "Address";
    request.approverFirstName = "FirstName";
    request.approverLastName = "LastName";
    request.approverPhone = "+12345678901";
    request.approverEmail = "email@domain.com";
    request.zipcode = "123456";
    request.city = "City";
    request.company = "Company";
    request.subjectAlternativeNames.push("test1.domain.com");
    request.subjectAlternativeNames.push("test2.domain.com");

    client.request.send(request,function(err, result){
        console.log(result);
        console.log('Error: '+ err);
        console.log('Errors: ');
        console.log(result.errors)
    });

}).call(this);

